platform/ext/target/nordic_nrf/common/cmsis_drivers/Driver_USART.c

```diff
@@ -0,0 +1,341 @@
+    nrfx_uarte_t        uarte;
+    nrfx_uarte_config_t config;
+    size_t              tx_count;
+    size_t              rx_count;
+} UARTx_Resources;
+
+//
+//   Functions
+//
+
+ARM_DRIVER_VERSION ARM_USART_GetVersion(void)
+{
+    return DriverVersion;
+}
+
+ARM_USART_CAPABILITIES ARM_USART_GetCapabilities(void)
+{
+    return DriverCapabilities;
+}
+
+int32_t ARM_USARTx_Initialize(ARM_USART_SignalEvent_t cb_event,
+                              UARTx_Resources *uart_resources)
+{
+    ARG_UNUSED(cb_event);
+
+    nrfx_err_t err_code = nrfx_uarte_init(&uart_resources->uarte,
+                                          &uart_resources->config,
+                                          NULL);
+    return (err_code == NRFX_SUCCESS) ? ARM_DRIVER_OK
+                                      : ARM_DRIVER_ERROR_BUSY;
+}
+
+int32_t ARM_USARTx_Uninitialize(UARTx_Resources *uart_resources)
+{
+    nrfx_uarte_uninit(&uart_resources->uarte);
+    return ARM_DRIVER_OK;
+}
+
+int32_t ARM_USARTx_PowerControl(ARM_POWER_STATE state,
+                                UARTx_Resources *uart_resources)
+{
+    ARG_UNUSED(uart_resources);
+
+    switch (state) {
+    case ARM_POWER_OFF:
+    case ARM_POWER_LOW:
+        return ARM_DRIVER_ERROR_UNSUPPORTED;
+    case ARM_POWER_FULL:
+        /* Nothing to be done */
+        return ARM_DRIVER_OK;
+    /* default:  The default is not defined intentionally to force the
+     *           compiler to check that all the enumeration values are
+     *           covered in the switch.*/
+    }
+}
+
+int32_t ARM_USARTx_Send(const void *data, uint32_t num,
+                        UARTx_Resources *uart_resources)
+{
+    nrfx_err_t err_code = nrfx_uarte_tx(&uart_resources->uarte, data, num);
+    if (err_code != NRFX_SUCCESS) {
+        return ARM_DRIVER_ERROR_BUSY;
+    }
+
+    uart_resources->tx_count = num;
+    return NRFX_SUCCESS;
+}
+
+int32_t ARM_USARTx_Receive(void *data, uint32_t num,
+                           UARTx_Resources *uart_resources)
+{
+    nrfx_err_t err_code = nrfx_uarte_rx(&uart_resources->uarte, data, num);
+    if (err_code != NRFX_SUCCESS) {
+        return ARM_DRIVER_ERROR_BUSY;
+    }
+
+    uart_resources->rx_count = num;
+    return NRFX_SUCCESS;
+}
+
+int32_t ARM_USART_Transfer(const void *data_out, void *data_in, uint32_t num)
+{
+    ARG_UNUSED(data_out);
+    ARG_UNUSED(data_in);
+    ARG_UNUSED(num);
+
+    return ARM_DRIVER_ERROR_UNSUPPORTED;
+}
+
+uint32_t ARM_USARTx_GetTxCount(const UARTx_Resources *uart_resources)
+{
+    return uart_resources->tx_count;
+}
+
+uint32_t ARM_USARTx_GetRxCount(const UARTx_Resources *uart_resources)
+{
+    return uart_resources->rx_count;
+}
+
+int32_t ARM_USARTx_Control(uint32_t control, uint32_t arg,
+                           UARTx_Resources *uart_resources)
+{
+    if ((control & ARM_USART_CONTROL_Msk) != ARM_USART_MODE_ASYNCHRONOUS) {
+        return ARM_DRIVER_ERROR_UNSUPPORTED;
+    }
+
+    nrfx_uarte_config_t *config = &uart_resources->config;
```

Comment:
If the function exits early, this config will not match the actual configuration since the uninit/init didn't run. WIll that be a problem?

Reply:
I thought about it. It should not be, as it is actually not described how the function should behave when only some of requested changes are successful. But using a local copy of the `config` structure seems to be safer. I'll rework it.